src/cfnlint/rules/resources/s3/BucketPolicy.py

```diff
@@ -0,0 +1,181 @@
+        if not isinstance(actions, (list, six.string_types)):
+            return matches
+        resources = value.get('Resource')
+        if not isinstance(resources, (list, six.string_types, dict)):
+            return matches
+
+        if isinstance(actions, six.string_types):
+            actions = [actions]
+        if isinstance(resources, (six.string_types, dict)):
+            resources = [resources]
+
+        found_object_actions, found_bucket_actions, actions_results_conclusive = self._get_statement_actions(actions)
+        found_object_resources, found_bucket_resources, bucket_results_conclusive = self._get_statement_resources(
+            resources, bucket_resources, custom_resources)
+        if not actions_results_conclusive or not bucket_results_conclusive:
+            return matches
+
+        if (not found_object_resources) and found_object_actions:
+            message = 'S3 Bucket Policy has object actions and no specified object resource'
+            matches.append(RuleMatch(path[:] + ['Resource'], message))
+        elif not found_bucket_resources and found_bucket_actions:
+            message = 'S3 Bucket Policy has bucket actions and no specified bucket resource'
+            matches.append(RuleMatch(path[:] + ['Resource'], message))
+
+        return matches
+
+    def _get_statement_actions(self, actions):
+        """ Breaks down the actions in a statement to each category """
+        # print(self.s3_actions)
+        object_actions = list(filter(re.compile('.*Object.*').match, self.s3_actions))
+        object_actions.extend(self.s3_object_actions)
+
+        bucket_actions = ['s3:' + action for action in list(set(self.s3_actions) - set(object_actions))]
+        object_actions = ['s3:' + action for action in object_actions]
+
+        found_object_actions = []
+        found_bucket_actions = []
+        for action in actions:
+            if action in object_actions:
+                found_object_actions.append(action)
+            elif action in bucket_actions:
+                found_bucket_actions.append(action)
+            else:
+                return None, None, False
+
+        return found_object_actions, found_bucket_actions, True
+
+    def _get_statement_resources(self, resources, bucket_resources, custom_resources):
+        """ Get Statement Resources """
+        found_bucket_resources = False
+        found_object_resources = False
+        for resource in resources:
+            if isinstance(resource, dict):
+                if len(resource) == 1:
+                    for func, func_value in resource.items():
+                        if func == 'Fn::Sub':
+                            if isinstance(func_value, six.string_types):
+                                if re.match('.+/.+', func_value):  # pylint: disable=R1703
+                                    found_object_resources = True
+                                else:
+                                    found_bucket_resources = True
```

Comment:
In this case, a `Resource: *` is counted as a "Bucket Resource"... 

Semantic not really true, is the the intended behaviour? Maybe treat `*` specifically or different?

Reply:
Trying to think of the different scenarios here.  Since this is in a sub it gets really interesting.
- `Resource: !Sub "*"`
- `Resource: !Sub "arn:aws:s3:::${myBucket}"` where myBucket is a Resource
- `Resource: !Sub "arn:aws:s3:::${myBucket}"` where myBucket is a parameter (this could be any string... we should skip this?)
- `Resource: !Sub "arn:aws:s3:::*"`

Maybe I can't safely determine what is going on here to know what to do.  I think there is a possibility if I know myBucket is a resource that I can still make the appropriate determination.  Thoughts?